### PR TITLE
feat(entities): add generic sort options

### PR DESCRIPTION
**Pull request by jeabakker**

### CODE REVIEW

File: views/json/resources/livesearch/groups.php [hunk 1]
```diff
@@ -12,8 +12,11 @@
 	'query' => $query,
 	'type' => 'group',
 	'limit' => $limit,
-	'sort' => 'name',
-	'order' => 'ASC',
+	'sort_by' => [[
+		'property_type' => 'metadata',
+		'property_value' => 'name',
```
Comment (jdalsem) [inline views/json/resources/livesearch/groups.php:17]: property

**Review by jdalsem — CHANGES_REQUESTED** [review]

File: docs/guides/search.rst [hunk 2]
```diff
@@ -36,11 +36,11 @@ In addition to all parameters accepted by ``elgg_get_entities()``, ``elgg_search
             'metadata' => ['description'],
             'annotations' => ['location'],
         ],
-        'sort' => [
+        'sort_by' => [[
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply docs/guides/search.rst:39]: added

File: engine/classes/Elgg/Menus/FilterSortItems.php [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	 *
+	 * @param \Elgg\Hook $hook 'register', 'menu:filter<:some filter_id>'
+	 *
+	 * @return MenuItems|null
+	 */
+	public static function registerTimeCreatedSorting(\Elgg\Hook $hook): ?MenuItems {
+		
+		if (!(bool) $hook->getParam('filter_sorting', true)) {
+			// sorting is disabled for this menu
+			return null;
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Menus/FilterSortItems.php:25]: the new standard is to have typehinted return values on all functions

File: engine/classes/Elgg/Menus/FilterSortItems.php [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+		
+		/* @var $result MenuItems */
+		$result = $hook->getValue();
+		
+		$result[] = \ElggMenuItem::factory([
+			'name' => 'sort:time_created:desc',
+			'icon' => 'sort-numeric-down-alt',
+			'text' => elgg_echo('sort:newest'),
+			'href' => elgg_http_add_url_query_elements(current_page_url(), [
+				'sort_by' => [[
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Menus/FilterSortItems.php:36]: fixed

File: engine/classes/Elgg/Menus/Page.php [hunk 5]
```diff
@@ -44,14 +44,21 @@ public static function registerAdminAdminister(\Elgg\Hook $hook) {
...
 			'name' => 'users',
 			'text' => elgg_echo('admin:users'),
 			'href' => 'admin/users',
 			'priority' => 40,
 			'section' => 'administer',
 		]);
+		$return[] = \ElggMenuItem::factory([
+			'name' => 'users2',
+			'text' => elgg_echo('admin:users'),
+			'href' => 'admin/usersasudfhgaisdgfiau',
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Menus/Page.php:57]: oops ;)

File: engine/hooks.php [hunk 6]
```diff
@@ -116,6 +116,9 @@
 		'menu:admin:users:bulk' => [
 			'Elgg\Menus\AdminUsersBulk::disableItems' => [],
 		],
+		'menu:page' => [
+			'Elgg\Menus\Page::prepareAdminAdminister' => [],
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/hooks.php:120]: changed

File: engine/lib/entities.php [hunk 7]
```diff
@@ -405,6 +408,26 @@ function elgg_get_site_entity() {
...
+ * 			'property_value' => 'time_created',
+ * 			'direction' => 'ASC',
+ * 		],
+ * 		[
+ * 			'property_type' => 'metadata',
+ * 			'property_value' => 'name',
+ * 			'direction' => 'DESC',
+ * 		],
+ * 		[
+ * 			'property_type' => 'private_setting',
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/lib/entities.php:425]: yep ;)

File: mod/groups/classes/Elgg/Groups/Menus/FilterSortItems.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+class FilterSortItems {
+	
+	/**
+	 * Register sorting menu items based on the relationship 'member'
+	 *
+	 * @param \Elgg\Hook $hook 'register', 'menu:filter<:some filter_id>'
+	 *
+	 * @return MenuItems|null
+	 */
+	public static function registerMembershipSorting(\Elgg\Hook $hook): ?MenuItems {
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply mod/groups/classes/Elgg/Groups/Menus/FilterSortItems.php:21]: keeping this inline with the core FilterSortItems

File: mod/search/classes/Elgg/Search/Search.php [hunk 9]
```diff
@@ -156,35 +156,59 @@ protected function initParams(array $params = []) {
...
-		$order = get_input('order', $default_order);
-		
 		$current_params = [
 			'query' => $query,
 			'offset' => $offset,
 			'limit' => $limit,
-			'sort' => $sort,
-			'order' => $order,
+			'sort_by' => $sort_by,
+			'sort' => $sort, // deprecated
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply mod/search/classes/Elgg/Search/Search.php:210]: Is handled in the ``Elgg\Search\SearchService``

File: views/json/resources/livesearch/groups.php (hunk 1 above)
Comment (jeabakker) [reply views/json/resources/livesearch/groups.php:17]: fixed

File: views/json/resources/livesearch/objects.php [hunk 10]
```diff
@@ -11,8 +11,11 @@
 	'type' => 'object',
 	'subtype' => elgg_extract('subtype', $vars),
 	'limit' => $limit,
-	'sort' => 'title',
-	'order' => 'ASC',
+	'sort_by' => [[
+		'property_type' => 'metadata',
+		'property_value' => 'title',
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply views/json/resources/livesearch/objects.php:16]: fixed

File: views/json/resources/livesearch/users.php [hunk 11]
```diff
@@ -12,8 +12,11 @@
 	'query' => $query,
 	'type' => 'user',
 	'limit' => $limit,
-	'sort' => 'name',
-	'order' => 'ASC',
+	'sort_by' => [[
+		'property_type' => 'metadata',
+		'property_value' => 'name',
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply views/json/resources/livesearch/users.php:17]: done and fixed

File: engine/classes/Elgg/Database/Clauses/EntitySortByClause.php [hunk 12]
```diff
@@ -48,36 +48,63 @@ public function prepare(QueryBuilder $qb, $table_alias = null) {
...
-				$md_alias = $qb->joinMetadataTable($table_alias, 'guid', $this->property, $this->join_type);
-				$column = "$md_alias.value";
+			case 'metadata':
+				$md_alias = $qb->joinMetadataTable($table_alias, $from_column, $this->property, $this->join_type);
+				$column = "{$md_alias}.value";
 				break;
 
-			case 'attribute' :
+			case 'attribute':
 				if (!in_array($this->property, ElggEntity::PRIMARY_ATTR_NAMES)) {
```
[the remark this answers is not in the document]
Comment (jeabakker) [reply engine/classes/Elgg/Database/Clauses/EntitySortByClause.php:69]: agree, but not changing this. It's documented that not all sort features might work on non ``elgg_get_entities`` functions

**Review by jdalsem — APPROVED** [review]

**Review by jdalsem — APPROVED** [review]

**Pull request merged**